double/double.man
.\" -*- nroff -*-
.TH DOUBLE 1  "22 February 1989"
.SH NAME
double \- filter for converting and editing double precision data
.SH SYNOPSIS
.B double
[\-\fBoption\fR [\fIvalue\fR]] ...
.SH DESCRIPTION
.IX double \- filter for converting, scaling and cutting double precision data
.I double
copies double precision data from the standard input to the standard output.
The user can specify optional scaling, editing and merging of additional data
from specified files.
.PP
If you specify more than one operation on the command line, they will be
applied in the following order.  See `OPTIONS' for a more complete description
of these.  The specified
.I offset
number of values are skipped.
While there is still data on the standard input one value is read.
One value from the
.I add
file is read and added to the value if present.
One value from the
.I mult
file is read and the value is multiplied by it if present.
The value is multiplied by the
.I scale
value if present.  The
.I baseline
value, if present, is added to the value.
The value is writen to the standard output.
One value is read from the
.I join
file if present and written to the standard output.
Last,
.I count
values in the standard input are skipped.
.PP
The user can achieve an arbitrary order by implementing each operation as
a single invocation of
.I double
in a multiple stage pipline.
.SH OPTIONS
.TP 20
.BI bi
the input contains binary (unformatted double precision) data.
.TP 
.BI bo
the output will contain binary (unformatted double precision) data.
.TP 
.BI "add FILE"
add the next input value from
.I FILE
to each output value.
.TP 
.BI "baseline DOUBLE"
add
.I  DOUBLE
to each output value (default 0.0).
.TP 
.BI "copy INTEGER"
copy 
.I INTEGER
values to the output each time around (default 1).
.TP 
.BI "join FILE"
 merge one value from
.I FILE
into output with each other output value.
.TP 
.BI "mult FILE"
file - multiply each output value with the next value from
.I FILE.
.TP 
.BI "offset INTEGER"
skip
.I  INTEGER
values from the beginning of the the input file before
beginning to copy (default 0).
.TP 
.BI "scale DOUBLE"
scale every output value by
.I  DOUBLE
(default 1.0).
.TP 
.BI "skip INTEGER"
skip
.I INTEGER
values after copying each cycle(default 0).
.SH EXAMPLES
.nf

      vector scalar product bvec[i] = 5.0 * avec[i]:
      double -scale 5.0 <avec >bvec

      vector addition cvec[i] = avec[i] + bvec[i]:
      double -add avec <bvec >cvec

      vector multiplication cvec[i] = avec[i] * bvec[i]:
      double -mult avec <bvec >cvec
      
      vector editing:
      bvec[i] = avec[2i]
      double -skip 1 <avec >bvec

      bvec[i] = avec[i+5]
      double -offset 5 <avec >bvec

      bvec[i] = avec[2i]
      bvec[i+1] = avec[2i+1]
      double -copy 2 -skip 2 <avec >bvec

.fi
.SH "SEE ALSO"
csh(1), sh(1)
.SH BUGS
This is virtually untested and still undergoing developmental changes.
Please mail any bugs reports or comments to rich@rice.edu, and I'll
provide support if I can.


ChangeLog
Wed Mar 27 12:42:25 1991  Richard Murphey  (rich at tahomah)

	* graphics.texinfo: added a note about using `which graph' to
	check whether /bin/graph is being mistakenly used.  Removed
	references to `unix' where unnecessary.

Fri Mar 15 15:12:11 1991  Richard Murphey  (rich at tahomah)

	* graphics.texinfo: Tektronix was spelled improperly.

Tue Mar 12 01:30:15 1991  Richard Murphey  (rich at kappa)

	* regresstion tests on sun3, sun4 and sequent pass. This si
	version 0.16. double 0.13 graph 0.15 plot2fig 0.4 plot2plot 0.1
	plot2ps 0.4 plot2tek 0.3 tek2plot 0.1 xplot 0.4

Tue Mar  5 14:58:25 1991  Richard Murphey  (rich at kappa)

	* graphics.texinfo: added more examples for graph.
	graph.c: added warning if margins are to small for labels.
	version 0.15: double 0.13 graph 0.15 plot2fig 0.3 plot2plot 0.1
	plot2ps 0.3 plot2tek 0.2 tek2plot 0.1 xplot 0.3

Sun Mar  3 12:22:42 1991  Richard Murphey  (rich at kappa)

	* added byte ordering code to allow for a default host dependent
	byte ordering.  This should be more compatible with byte ordering
	of plot files in other versions of the plot utilities. This is
	version 0.14:
	double 0.13 graph 0.14 plot2fig 0.3 plot2plot 0.1 plot2ps 0.3
	plot2tek 0.2 tek2plot 0.1 xplot 0.3

	* We were announced on comp.archives today, so this is a good
	time to clean up and release: version 0.13:
	double 0.13 graph 0.14 plot2fig 0.3 plot2plot 0.1 plot2ps 0.3
	plot2tek 0.2 tek2plot 0.1 xplot 0.2

Sat Feb  9 15:01:20 1991  Rich Murphey  (murphey at cns15)

	* graphics.texinfo: added description of -N and -R options.

Thu Nov  8 00:09:57 1990  Richard Murphey  (rich at kappa)

	* cleaned up regression tests and makefiles. Version 0.10

Wed Nov  7 11:31:05 1990  Richard Murphey  (rich at kappa)

	* graphics.texinfo: added Ray's description of long option names.
	* graph: more fixes to arguement parsing. This is version 0.9.

	* merged Ray's fixes for plto2tek, plot2plot, and graph as well
	as lots of new code.  This is version 0.8.

Tue Nov  6 13:52:06 1990  Ray Toy  (toy at dino.ecse.rpi.edu)

	* Added support for log axes in graph.c and write_plot.c.
	Doesn't quite work when margin != 0, though.  Added some
	info about it in graph.texinfo.

	* plot2plot modified to use GNU getopt.

Wed Oct 31 23:24:09 1990  Richard Murphey  (rich at tahomah)

	* wrote plot2plot which reads and writes plot files in both
	ascii and binary formats and provides backward compatibility.
	* Seperated the copying message out of each program and had
	each include it from the COPYING file. This is version 0.7.
	* graphics.texinfo: condensed documentation into one file.

Mon Oct 29 13:44:49 1990  Richard Murphey  (rich at tahomah)

	* added changes contributed by Ray Toy <toy@dino.ecse.rpi.edu>
	to graph.  This is version 0.5

Thu Oct 18 11:02:18 1990  Richard Murphey  (rich at tahomah)

	* added prototypes to all library routine declarations and
	removed lint from much of the code. This is version 0.3.

	* Makefile: added Ultrix Vax target from Jeff Templon.




double/double.c
/* double - filter for converting, scaling and cutting unformatted (binary) or
   ascii double precision data streams.

   Copyright (C) 1989 Free Software Foundation, Inc.

   Plot is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY.  No author or distributor accepts responsibility to
   anyone for the consequences of using it or for whether it serves any
   particular purpose or works at all, unless he says so in writing.
   Refer to the GNU General Public License for full details.

   Everyone is granted permission to copy, modify and redistribute plot,
   but only under the conditions described in the GNU General Public
   License.  A copy of this license is supposed to have been given to you
   along with plot so you can know your rights and responsibilities.  It
   should be in a file named COPYING.  Among other things, the copyright
   notice and this notice must be preserved on all copies.  */

#include <stdio.h>
#include <ctype.h>
#include <math.h>

#ifdef STDC_HEADERS
#include <string.h>
#include <stdlib.h>
#else
#ifdef	USG
#include <string.h>
#include <malloc.h>
#else
#include <strings.h>
#include <malloc.h>
#endif	/* USG */
#endif	/* STDC_HEADERS */

#include <sys/file.h>

char *help_message="\n\
double - a utility for converting, scaling, cutting and pasting unformatted\n\
 (binary) or ascii double precision data.\n\
\n\
Usage:                 double [-option [ARG]] ...<DATA\n\
\n\
Option:         Description:\n\
\n\
add FILE        add values from FILE to each value from DATA.\n\
baseline VALUE  add VALUE to each value from DATA (default 0.0).\n\
bi              read binary (unformatted double precision) values from DATA.\n\
bo              output binary (unformatted double precision) data.\n\
fo              output binary (unformatted single precision) data.\n\
copy NUMBER     output NUMBER values before skipping (default 1).\n\
h               print this help message\n\
join FILE       output one value from FILE for each value from DATA.\n\
mult FILE       multiply values from FILE with each value from DATA.\n\
max             print out the maximum data value.\n\
min             print out the minimum data value.\n\
offset NUMBER   ignore NUMBER values before reading value from DATA.\n\
scale VALUE     scale every output datum by VALUE (default 1.0).\n\
skip NUMBER     skip NUMBER values from DATA after each copying (default 0).\n\
silent        	don't output any of the data values.\n\
si              the input contains binary short integers.\n\
sw              reverse the bytes of the short integers on input.\n\
\n\
      Example:\n\
      vector scalar product bvec[i] = 5.0 * avec[i]:\n\
      double -scale 5.0 <avec >bvec\n\
\n\
      vector addition cvec[i] = avec[i] + bvec[i]:\n\
      double -add avec <bvec >cvec\n\
\n\
      vector multiplication cvec[i] = avec[i] * bvec[i]:\n\
      double -mult avec <bvec >cvec\n\
      \n\
      vector editing:\n\
      bvec[i] = avec[2i]\n\
      double -skip 1 <avec >bvec\n\
\n\
      bvec[i] = avec[i+5]\n\
      double -offset 5 <avec >bvec\n\
\n\
      bvec[i] = avec[2i]\n\
      bvec[i+1] = avec[2i+1]\n\
      double -copy 2 -skip 2 <avec >bvec\n\
\n\
      maximum and minimum elements of a vector:\n\
      double -max -silent <avec\n\
      double -min -silent <avec\n\
      rich@rice.edu\n\
\n\
";

int binary_output=0, float_output=0, binary_input=0;
int short_int_input=0, swap_bytes=0;
int take_derivative=0;
#define LARGE (9e30)
double last_data = LARGE;

/* readit and writeit return 1 on success and 0 on error */
int
readit (data, fp)
     double *data;
     FILE *fp;
{
  int ret;
  if (binary_input)
    {
      /* returns the number of bytes actually read or 0 for end-of-file */
      ret = read( fileno(fp), (char*)data, sizeof (double));
      return (ret > 0);
    }
  else if (short_int_input)
    {
      /* returns the number of bytes actually read or 0 for end-of-file */
      short int a;
      ret = read( fileno(fp), (char*)&a, sizeof (a));
      if (swap_bytes)
	*data = (short int)((a & 0xFF)<<8 | ((a>>8) & 0xFF));
      else
	*data = a;
      return (ret > 0);
    }
  else
    {
      /* returns EOF on error, zero if succesful */
      ret = fscanf (fp, "%lf", data);
      return (ret > 0);
    }
}

int
writeit  (data, fp)
     double *data;
     FILE *fp;
{
  int success = 0;
  if (binary_output)
    {
      /* number of bytes writen is returned.   -1 for error */
      success = (0 < write ( fileno(fp), (char*)data, sizeof (double)));
      return success;
    }
  else if (float_output)
    {
      float tmp;
      tmp = *data;
      /* number of bytes writen is returned.   -1 for error */
      success = (0 < write ( fileno(fp), (char*) &tmp, sizeof (float)));
      return success;
    }
  else
    {
#ifdef V
      /* V system: EOF is returned upon end of input, zero if succesful */
      success = (0 == fprintf (fp, "%.19g\n", *data));
#endif
#ifdef sparc
      /* sparc os4.0: fprintf returns EOF on error, else number
	 of characters written */
      int ret = fprintf (fp, "%#.19g\n", *data);
      success = (0 < ret);
#endif
#ifdef myrias
      /* myrias: fprintf returns EOF on error, else number of bytes written. */
      success = (0 < fprintf (fp, "%#.19g\n", *data));
#endif
#ifdef sun3
      /* sun3: fprintf returns EOF on error, zero if succesful */
      success = (0 <= fprintf (fp, "%#.19g\n", *data));
#endif
#ifdef sequent
      /* sequent: fprintf returns EOF on error, zero if succesful */
      success = (0 <= fprintf (fp, "%#.19g\n", *data));
#endif
      /* return zero if write failed. */
      return success;
    }
}

int
main (argc, argv)
     int argc;
     char **argv;
{
  int copy=1, offset=0, skip=0;
  char *add_file=0, *mult_file=0, *join_file=0;
  double scale=1., baseline=0.;
  FILE *add_fp = NULL;
  FILE *mult_fp = NULL;
  FILE *join_fp = NULL;
  double data=0., max = -1.0e+38, min=1.0e+38;
  int read_ok=1, write_ok=1, do_min=0, do_max=0, silent=0;

  if (argc <= 1)
    {
      printf( "%s", help_message);
      exit (0);
    }
      
  argc--;
  argv++;
  while (argc > 0)
    {
      /*
       *            OPTIONS WHICH ARE SWITCHES (NO ARGUMENT):
       */
      if (!strcmp (*argv, "-bo"))
	/*  output is binary */
	{
	  argc--; argv++;
	  binary_output = 1;
	}
      else if (!strcmp (*argv, "-fo"))
	/*  output is binary */
	{
	  argc--; argv++;
	  float_output = 1;
	}
      else if (!strcmp (*argv, "-bi"))
	/*  input is binary */
	{
	  argc--; argv++;
	  binary_input = 1;
	}
      else if (!strcmp (*argv, "-deriv"))
	/* take the derivative of the input. */
	{
	  argc--; argv++;
	  take_derivative = 1;
	}
      else if (!strcmp (*argv, "-si"))
	/*  input is binary short ints */
	{
	  argc--; argv++;
	  short_int_input = 1;
	}
      else if (!strcmp (*argv, "-sw"))
	/*  swap bytes on input */
	{
	  argc--; argv++;
	  swap_bytes = 1;
	}
      else if (!strcmp (*argv, "-h"))
	{
	  printf( "%s", help_message);
	  exit (0);
	}
      else if (!strcmp (*argv, "-max"))
	/* output maximum value at the end. */
	{
	  do_max = 1;
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-min"))
	/* output minimum value at the end. */
	{
	  do_min = 1;
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-silent"))
	/* do not output the data values. */
	{
	  silent = 1;
	  argc--; argv++;
	}
      /*
       *               OPTIONS WHICH TAKE ARGUMENTS:
       */
      else if (argc < 2)
	{
	  fprintf (stderr, "No argument following \"%s\" switch\n", *argv);
	  exit (-1);
	}
      else if (!strcmp (*argv, "-add"))
	/*  add values from FILE. */
	{
	  argc--; argv++;
	  add_file = *argv;
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-baseline"))
	/*  add BASELINE to each output value. */
	{
	  argc--; argv++;
	  baseline = atof (*argv);
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-copy"))
	/*  number of values to copy to the output each time around. */
	{
	  argc--; argv++;
	  copy = atoi (*argv);
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-join"))
	/*  merge file. */
	{
	  argc--; argv++;
	  join_file = *argv;
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-mult"))
	/*  multiply each output value by one value from FILE. */
	{
	  argc--; argv++;
	  mult_file = *argv;
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-offset"))
	/*  number of data to skip from the beginning of the file. */
	{
	  argc--; argv++;
	  offset = atoi (*argv);
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-scale"))
	/*  scale each output value by command line value. */
	{
	  argc--; argv++;
	  scale = atof (*argv);
	  argc--; argv++;
	}
      else if (!strcmp (*argv, "-skip"))
	/* number of values to skip between those we copy. */
	{
	  argc--; argv++;
	  skip = atoi (*argv);
	  argc--; argv++;
	}
      else
	{
	  fprintf (stderr, "unknown switch \"%s\"\n", *argv);
	  argc--; argv++;
	}
    }
  
  /*  open additive file. */
  if (add_file )
    {
      add_fp = fopen (add_file, "r");
      if (add_fp == NULL)
	{
	  printf ("%s: unable to open: %s\n", argv[0], add_file);
	  exit (-1);
	}
    }
  
  /*  open multiplicative scaling file. */
  if (mult_file )
    {
      mult_fp = fopen (mult_file, "r");
      if (mult_fp == NULL)
	{
	  printf ("%s: unable to open: %s\n", argv[0], mult_file);
	  exit (-1);
	}
    }
  
  
  /*  open file of data to be joined in. */
  if (join_file )
    {
      join_fp = fopen (join_file, "r");
      if (join_fp == NULL)
	{
	  printf ("%s: unable to open: %s\n", argv[0], join_file);
	  exit (-1);
	}
    }
  
  
  /*  skip data upto the offset datum. */
  while (offset-- &&  read_ok)
    {
      read_ok = readit (&data, stdin);
    }
  
  
  while (read_ok )
    {
      int count = copy;
      double mult_data, add_data;
      while (count-- &&  read_ok)
	{
	  read_ok = readit (&data, stdin);
	  if ( ! read_ok )
	    goto finish;
	  if (add_fp != NULL)
	    {
	      read_ok = readit (&add_data, add_fp);
	      if ( ! read_ok )
		goto finish;
	      data += add_data;
	    }
	  if (mult_fp != NULL)
	    {
	      read_ok = readit (&mult_data, mult_fp);
	      if ( ! read_ok )
		goto finish;
	      data *= mult_data;
	    }
	  data *= scale;
	  data += baseline;
	  if (data > max)
	    max = data;
	  if (data < min)
	    min = data;
	  
	  if (take_derivative)
	    {
	      double tmp = data;
	      if (last_data == LARGE)
		data = last_data;
	      data -= last_data;
	      last_data = tmp;
	    }
	  if (! silent)
	    {
	      write_ok = writeit (&data, stdout);
	      if ( ! write_ok)
		{
		  fprintf ( stderr, "double: unable to write to output.\n");
		  exit (-1);
		}
	      if (join_fp != NULL)
		{
		  read_ok = readit (&data, join_fp);
		  if (read_ok)
		    write_ok = writeit (&data, stdout);
		  if ( ! write_ok)
		    {
		      fprintf ( stderr, "double: unable to write to output.\n");
		      exit (-1);
		    }
		}
	    }
	}
      count = skip;
      while (count-- &&  read_ok)
	{
	  read_ok = readit (&data, stdin);
	}
    }
 finish:
  if (do_max)
    writeit (&max, stdout);
  if (do_min)
    writeit (&min, stdout);
  return 0;
}
